x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/output/AutodetectResultProcessor.java

```diff
@@ -206,6 +212,27 @@ public void setProcessKilled() {
         renormalizer.shutdown();
     }
 
+    void handleOpenForecasts() {
+        try {
+            if (runningForecasts.isEmpty() == false) {
+                LOGGER.warn("[{}] still had forecasts {} executing. Attempting to set them to failed.",
+                    jobId,
+                    runningForecasts.keySet());
+                // There may be many docs in the results persistence queue. But we only want to bother updating the running forecasts
+                bulkResultsPersister.clearBulkRequest();
+                for (ForecastRequestStats forecastRequestStats : runningForecasts.values()) {
```

Comment:
It would be nice if we could reuse here `JobResultsProvider.setRunningForecastsToFailed`. The benefits would be that we don't need a copy constructor for `ForecastRequestStats` which is prone for bugs (forgetting to add a new field there), and we don't need to keep track of open forecasts here in the new `runningForecasts` member.

Reply:
`setRunningForecastsToFailed` is a DBQ. It is possible to have a forecast request stat document persisted without the index refreshing since. So, the DBQ would not find it.